server/src/main/java/org/opensearch/search/pipeline/Pipeline.java

```diff
@@ -117,92 +119,135 @@ protected void afterResponseProcessor(Processor processor, long timeInNanos) {}
-            beforeTransformRequest();
-            try {
-                try (BytesStreamOutput bytesStreamOutput = new BytesStreamOutput()) {
-                    request.writeTo(bytesStreamOutput);
-                    try (StreamInput in = bytesStreamOutput.bytes().streamInput()) {
-                        try (StreamInput input = new NamedWriteableAwareStreamInput(in, namedWriteableRegistry)) {
-                            request = new SearchRequest(input);
-                        }
-                    }
-                }
-                for (SearchRequestProcessor processor : searchRequestProcessors) {
-                    beforeRequestProcessor(processor);
-                    long start = relativeTimeSupplier.getAsLong();
-                    try {
-                        request = processor.processRequest(request);
-                    } catch (Exception e) {
-                        onRequestProcessorFailed(processor);
-                        if (processor.isIgnoreFailure()) {
-                            logger.warn(
-                                "The exception from request processor ["
-                                    + processor.getType()
-                                    + "] in the search pipeline ["
-                                    + id
-                                    + "] was ignored",
-                                e
-                            );
-                        } else {
-                            throw e;
-                        }
-                    } finally {
-                        long took = TimeUnit.NANOSECONDS.toMillis(relativeTimeSupplier.getAsLong() - start);
-                        afterRequestProcessor(processor, took);
-                    }
+    void transformRequest(SearchRequest request, ActionListener<SearchRequest> requestListener) throws SearchPipelineProcessingException {
+        if (searchRequestProcessors.isEmpty()) {
+            requestListener.onResponse(request);
+            return;
+        }
+
+        try (BytesStreamOutput bytesStreamOutput = new BytesStreamOutput()) {
+            request.writeTo(bytesStreamOutput);
+            try (StreamInput in = bytesStreamOutput.bytes().streamInput()) {
+                try (StreamInput input = new NamedWriteableAwareStreamInput(in, namedWriteableRegistry)) {
+                    request = new SearchRequest(input);
                 }
-            } catch (Exception e) {
-                onTransformRequestFailure();
-                throw new SearchPipelineProcessingException(e);
-            } finally {
-                long took = TimeUnit.NANOSECONDS.toMillis(relativeTimeSupplier.getAsLong() - pipelineStart);
-                afterTransformRequest(took);
             }
+        } catch (IOException e) {
+            requestListener.onFailure(new SearchPipelineProcessingException(e));
+            return;
+        }
+
+        long[] pipelineStart = new long[1];
```

Comment:
Shouldn't we assign the value here? 

```suggestion
        long[] pipelineStart = new long[1] { relativeTimeSupplier.getAsLong() };
```

Reply:
I think it will be easier this way and we could have it `final` as well, wdyt?

Reply:
I think assignment happens on line 186.

Reply:
It is but could it be moved up?

Reply:
But then it would be a few nano seconds too soon!  😄   I think the argument that you can make it final is a good one.

Also, I thought there was a particular reason Michael wanted it on heap (new long[]).

Reply:
> But then it would be a few nano seconds too soon! 

That was the reasoning. :)

The construction of the chain should be pretty fast, but it's technically outside of the time spent executing the pipeline. 

Or is it? We do have the short-circuit logic up top to avoid constructing the listener chain if there are no request processors. One could argue that the time spent constructing the listener chain is a cost of pipeline execution that should be tracked.

I'm happy to just assign this one directly.

Reply:
> I'm happy to just assign this one directly.

I would just do that to be fair, it make code much simpler and readable